boot/modeenv.go

```diff
@@ -134,46 +119,100 @@ func (m *Modeenv) WriteTo(rootdir string) error {
 	if m.Mode == "" {
 		return fmt.Errorf("internal error: mode is unset")
 	}
-	fmt.Fprintf(buf, "mode=%s\n", m.Mode)
-
-	if m.RecoverySystem != "" {
-		fmt.Fprintf(buf, "recovery_system=%s\n", m.RecoverySystem)
-	}
-	if len(m.CurrentRecoverySystems) != 0 {
-		// recovery system label is composed of letters, numbers and dashes
-		fmt.Fprintf(buf, "current_recovery_systems=%s\n", asModeenvStringList(m.CurrentRecoverySystems))
-	}
-	if m.Base != "" {
-		fmt.Fprintf(buf, "base=%s\n", m.Base)
-	}
-	if m.TryBase != "" {
-		fmt.Fprintf(buf, "try_base=%s\n", m.TryBase)
-	}
-	if m.BaseStatus != "" {
-		fmt.Fprintf(buf, "base_status=%s\n", m.BaseStatus)
-	}
-	if len(m.CurrentKernels) != 0 {
-		fmt.Fprintf(buf, "current_kernels=%s\n", asModeenvStringList(m.CurrentKernels))
-	}
+	marshalNonEmptyForModeenvEntry(buf, "mode", m.Mode)
+	marshalNonEmptyForModeenvEntry(buf, "recovery_system", m.RecoverySystem)
+	marshalNonEmptyForModeenvEntry(buf, "current_recovery_systems", m.CurrentRecoverySystems)
+	marshalNonEmptyForModeenvEntry(buf, "base", m.Base)
+	marshalNonEmptyForModeenvEntry(buf, "try_base", m.TryBase)
+	marshalNonEmptyForModeenvEntry(buf, "base_status", m.BaseStatus)
+	marshalNonEmptyForModeenvEntry(buf, "current_kernels", strings.Join(m.CurrentKernels, ","))
 	if m.Model != "" || m.Grade != "" {
 		if m.Model == "" {
 			return fmt.Errorf("internal error: model is unset")
 		}
 		if m.BrandID == "" {
 			return fmt.Errorf("internal error: brand is unset")
 		}
-		fmt.Fprintf(buf, "model=%s/%s\n", m.BrandID, m.Model)
-	}
-	if m.Grade != "" {
-		fmt.Fprintf(buf, "grade=%s\n", m.Grade)
+		marshalNonEmptyForModeenvEntry(buf, "model", &modeenvModel{brandID: m.BrandID, model: m.Model})
 	}
+	marshalNonEmptyForModeenvEntry(buf, "grade", m.Grade)
 
 	if err := osutil.AtomicWriteFile(modeenvPath, buf.Bytes(), 0644, 0); err != nil {
 		return err
 	}
 	return nil
 }
 
+type modeenvValueMarshaller interface {
+	MarshalModeenvValue() (string, error)
+}
+
+type modeenvValueUnmarshaller interface {
+	UnmarshalModeenvValue(value string) error
+}
+
+func marshalNonEmptyForModeenvEntry(out io.Writer, key string, what interface{}) error {
```

Comment:
maybe this should be marshalNonEmptyModeenvEntryTo ? also I understand why the NonEmpty but to me it makes it a bit unclear whether it means what can't be empty, but it means it will be omitted.  Maybe just marshalModeenvEntryTo ?
and then a comment before to explain empty entries are omitted?